collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java

```diff
@@ -729,47 +739,107 @@ protected void onPostExecute(HashMap<String, Object> results) {
-                    case "application/vnd.google-apps.folder":
-                        dirs.add(new DriveListItem(f.getName(), "", f.getModifiedTime(), "", "",
-                                DriveListItem.DIR, f.getId(), parentId));
-                        break;
-                    default:
-                        // skip the rest of the files
-                        break;
+    }
+
+    private void filterForms(HashMap<String, Object> pageDetails) {
+        List<com.google.api.services.drive.model.File> fileList =
+                (List<com.google.api.services.drive.model.File>) pageDetails.get(FILE_LIST_KEY);
+        String parentId = (String) pageDetails.get(PARENT_ID_KEY);
+        String currentDir = (String) pageDetails.get(CURRENT_ID_KEY);
+
+        List<DriveListItem> dirs = new ArrayList<>();
+        List<DriveListItem> forms = new ArrayList<>();
+
+        for (com.google.api.services.drive.model.File f : fileList) {
+            String type = f.getMimeType();
+            switch (type) {
+                case "application/xml":
+                case "text/xml":
+                case "application/xhtml":
+                case "text/xhtml":
+                case "application/xhtml+xml":
+                    forms.add(new DriveListItem(f.getName(), "", f.getModifiedTime(), "", "",
+                            DriveListItem.FILE, f.getId(), currentDir));
+                    break;
+                case "application/vnd.google-apps.folder":
+                    dirs.add(new DriveListItem(f.getName(), "", f.getModifiedTime(), "", "",
+                            DriveListItem.DIR, f.getId(), parentId));
+                    break;
+                default:
+                    // skip the rest of the files
+                    break;
+            }
+        }
+        Collections.sort(dirs);
+        Collections.sort(forms);
+        driveList.addAll(dirs);
+        driveList.addAll(forms);
+        checkFormUpdates();
+    }
+
+    private void checkFormUpdates() {
+        for (DriveListItem item: driveList) {
+            if (item.getType() == DriveListItem.FILE) {
+                if (isNewerFormVersionAvailable(item) || areNewerMediaFilesAvailable(item)) {
+                    item.setNewerVersion(true);
                 }
             }
-            Collections.sort(dirs);
-            Collections.sort(forms);
-            driveList.addAll(dirs);
-            driveList.addAll(forms);
-            updateAdapter();
         }
     }
 
+    private boolean isNewerFormVersionAvailable(DriveListItem item) {
+        FormsDao formsDao = new FormsDao();
+        Cursor cursor = formsDao.getFormsCursorForFormFilePath(new StoragePathProvider().getDirPath(StorageSubdirectory.FORMS) + File.separator + item.getName());
+        if (cursor != null && cursor.moveToFirst()) {
+            Long lastModifiedLocal = new File(new StoragePathProvider().getDirPath(StorageSubdirectory.FORMS) + File.separator + item.getName()).lastModified();
```

Comment:
I think it;s time to use injected `StoragePathProvider` because in many places we use `new StoragePathProvider()`. It's already available in [AppDependencyModule](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java#L256) and this activity is also ready to handle injected objects https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java#L127 so you can just use it like we do with [GoogleAccountsManager](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java#L110) for example.